### PR TITLE
Adding the ability to process numbers in text files.

### LINKED ISSUE #993: [rook] relative error in XMLdiff
--------
Issue Description
--------
##### What did you expect to see happen?
When specifying a relative error, the acceptable diff between the test and gold values should be scaled on the gold value itself. For example gold=3.14e6 and test=3.15e6 should be acceptable to a `rel_err` of 1e-1.

##### What did you see instead?
The absolute difference is used instead.

##### Do you have a suggested fix for the development team?
A common numerical diffing algorithm could be used by all the `rook` differs.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
This lets things like relative error be handled.

Also, changing XMLDiff to use relative not absolute error as the
documentation states.

--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #993 

##### What are the significant changes in functionality due to this change request?
Changes XMLDiff to use relative error.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by PaulTalbot-INL** [comment]
> Job [Test qsubs](https://civet.inl.gov/job/339058/) on [970d769](https://github.com/idaholab/raven/commit/970d769955f9014ff531afd5f3afbcc08f95d6cc) : invalidated by @joshua-cogliati-inl
> 
> nothing should fail at fetch and branch

It does if `git` is unrecognized...

I'm trying to find all the pieces that went missing currently. So far it's `git`, `pdflatex`, and a version issue for `gcc`.

**Comment by joshua-cogliati-inl** [comment]
"collect2: fatal error: cannot find 'ld'"  

Hm, I would have to say that a "gcc" module that doesn't include ld probably needs to be fixed since you cannot compile anything with it.

**Comment by PaulTalbot-INL** [comment]
I tried to add "git" but the default module loads a ton of other modules. use.moose has a clean git we can use; I'm testing that set of libraries now on falcon.

**Comment by PaulTalbot-INL** [comment]
I think the failing RAVEN code (run by RAVEN) tests are likely due to the IO of the "inner" runs. What do you think of changing the verbosity to "quiet" on the "inner" input files? It makes debugging a little harder, but might make the tests much more timely.

**Comment by joshua-cogliati-inl** [comment]
I switched the verbosity from debug to all for the two Raven Running Raven tests.

**Comment by wangcj05** [comment]
@joshua-cogliati-inl could you resolve the conflicts?

**Comment by joshua-cogliati-inl** [comment]
@wangcj05 I fixed the conflict and the tests pass.

File: rook/NumTextDiff.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    self.__text_opts['remove_whitespace'] = bool(self.specs['remove_whitespace'])
+
+  def check_output(self):
+    """
+      Checks that the output matches the gold.
+      returns (same, message) where same is true if the
+      test passes, or false if the test failes.  message should
+      gives a human readable explaination of the differences.
+      @ In, None
+      @ Out, (same, message), same is true if the tests passes.
```
Comment (wangcj05) [inline rook/NumTextDiff.py]: types for "same, message"

File: rook/XMLDiff.py [hunk 2]
```diff
@@ -18,9 +18,12 @@
 import sys
 import os
 import xml.etree.ElementTree as ET
+from Tester import Differ
 import DiffUtils as DU
+#CSWF Defined because otherwise lines of code get too long.
+CSWF = DU.compare_strings_with_floats
```
Comment (wangcj05) [inline rook/XMLDiff.py]: lower case for "CSWF"? consistent with previous changes?

**Comment by wangcj05** [comment]
@joshua-cogliati-inl, I have some minor comments for this PR.

**Comment by alfoa** [comment]
@joshua-cogliati-inl

File: rook/NumTextDiff.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply rook/NumTextDiff.py]: Added types.

File: rook/XMLDiff.py (hunk 2 above)
Comment (joshua-cogliati-inl) [reply rook/XMLDiff.py]: changed to cswf

**Comment by alfoa** [comment]
Checklist passed...

PR can be approved and merged.

**Pull request merged**